### PR TITLE
[AutoDiff] Move ActivityAnalysis to a separate file.

### PR DESCRIPTION
**Pull request by bgogul**
Cleanup as part of https://bugs.swift.org/browse/TF-993

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: include/swift/SILOptimizer/Utils/Differentiation.h [hunk 1]
```diff
@@ -17,12 +17,12 @@
 // NOTE: Although the AD feature is developed as part of the Swift for
 // TensorFlow project, it is completely independent from TensorFlow support.
 //
-// TODO: Move definitions here from Differentiation.cpp.
+// TODO: Move definitions from lib/SILOptimizer/Mandatory/Differentiation.cpp.
```
Comment (dan-zheng) [inline include/swift/SILOptimizer/Utils/Differentiation.h]: I'm curious about what this TODO comment means. Do you plan to do sth like move definitions to a new file `lib/SILOptimizer/Utils/Differentiation.cpp`?

**Review by dan-zheng — APPROVED** [review]
Nice! I pushed some minor formatting changes, hope that's okay.

File: include/swift/SILOptimizer/Utils/Differentiation.h (hunk 1 above)
Comment (bgogul) [reply include/swift/SILOptimizer/Utils/Differentiation.h]: This comment was already there. So, I just left it there. I presume that this is related to moving some helper functions out of Differentiation.cpp? @rxwei might have more context.

**Comment by bgogul** [comment]
> Nice! I pushed some minor formatting changes, hope that's okay.

Of course!

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Comment by dan-zheng** [comment]
I'm done pushing changes now, sorry for not communicating that clearly 🙂

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: include/swift/SILOptimizer/Utils/Differentiation.h (hunk 1 above)
Comment (dan-zheng) [reply include/swift/SILOptimizer/Utils/Differentiation.h]: Oh, I understand now - the comment existed previously in `lib/SILOptimizer/Mandatory/Differentiation.h`.

File: include/swift/SILOptimizer/Analysis/ActivityAnalysis.h [hunk 2]
```diff
@@ -0,0 +1,235 @@
+//===--- ActivityAnalysis.h - SIL Activity Analysis -------------*- C++ -*-===//
```
Comment (rxwei) [inline include/swift/SILOptimizer/Analysis/ActivityAnalysis.h]: SIL Differentiable Activity Analysis.

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Comment by bgogul** [comment]
@swift-ci please clean test tensorflow

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Analysis/CMakeLists.txt [hunk 3]
```diff
@@ -2,9 +2,9 @@ silopt_register_sources(
   ARCAnalysis.cpp
   AccessSummaryAnalysis.cpp
   AccessedStorageAnalysis.cpp
-  // SWIFT_ENABLE_TENSORFLOW
+  # SWIFT_ENABLE_TENSORFLOW
```
Comment (dan-zheng) [inline lib/SILOptimizer/Analysis/CMakeLists.txt]: Oops, sorry about the wrong syntax!

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: include/swift/SILOptimizer/Analysis/ActivityAnalysis.h [hunk 4]
```diff
@@ -0,0 +1,235 @@
+//===--- ActivityAnalysis.h - Differentiable Activity Analysis --*- C++ -*-===//
```
Comment (rxwei) [inline include/swift/SILOptimizer/Analysis/ActivityAnalysis.h]: I think this file should be called `DifferentiableActivityAnalysis`. "Activity analysis" is not well-defined out of AD.

File: include/swift/SILOptimizer/Utils/Differentiation.h [hunk 5]
```diff
@@ -35,6 +37,37 @@ using llvm::SmallDenseSet;
...
+
+//===----------------------------------------------------------------------===//
+// Helpers
+//===----------------------------------------------------------------------===//
+
+namespace autodiff {
+
+/// Prints an "[AD] " prefix to `llvm::dbgs()` and returns the debug stream.
+/// This is being used to print short debug messages within the AD pass.
+raw_ostream &getADDebugStream();
```
Comment (rxwei) [inline include/swift/SILOptimizer/Utils/Differentiation.h]: Where is the implementation of these functions? When we define a `Utils/Differentiation.h`, we should make sure to have a `Utils/Differentiation.cpp` that implements all functions the header file declares.

Comment (bgogul) [reply include/swift/SILOptimizer/Utils/Differentiation.h]: They were in `SILOptimizer/Mandatory/Differentiation.cpp`. I created a`Utils/Differentiation.cpp`.

File: include/swift/SILOptimizer/Analysis/ActivityAnalysis.h (hunk 4 above)
Comment (bgogul) [reply include/swift/SILOptimizer/Analysis/ActivityAnalysis.h]: Done.

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: include/swift/SILOptimizer/Analysis/DifferentiableActivityAnalysis.h [hunk 6]
```diff
@@ -0,0 +1,235 @@
+//===--- ActivityAnalysis.h - Differentiable Activity Analysis --*- C++ -*-===//
```
Comment (dan-zheng) [inline include/swift/SILOptimizer/Analysis/DifferentiableActivityAnalysis.h]:
```suggestion
//===--- DifferentiableActivityAnalysis.h ---------------------*- C++ -*---===//
```

File: lib/SILOptimizer/Utils/Differentiation.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
+//===--- Differentiation.h - SIL Automatic Differentiation ----*- C++ -*---===//
```
Comment (dan-zheng) [inline lib/SILOptimizer/Utils/Differentiation.cpp]:
```suggestion
//===--- Differentiation.cpp - SIL Automatic Differentiation --*- C++ -*---===//
```

**Review by dan-zheng — APPROVED** [review]

**Pull request merged**